services/galley/src/Galley/Data.hs

```diff
@@ -934,100 +936,100 @@ addLocalMembersToRemoteConv qconv users = do
-      setType BatchUnLogged
-      setConsistency Quorum
-      for_ (mupOtrMuteStatus mup) $ \ms ->
-        addPrepQuery
-          Cql.updateOtrMemberMutedStatus
-          (ms, mupOtrMuteRef mup, lUnqualified lcid, lUnqualified luid)
-      for_ (mupOtrArchive mup) $ \a ->
-        addPrepQuery
-          Cql.updateOtrMemberArchived
-          (a, mupOtrArchiveRef mup, lUnqualified lcid, lUnqualified luid)
-      for_ (mupHidden mup) $ \h ->
-        addPrepQuery
-          Cql.updateMemberHidden
-          (h, mupHiddenRef mup, lUnqualified lcid, lUnqualified luid)
-    pure (memberUpdateToData (unTagged luid) mup)
-  updateMemberRemoteConv (Tagged (Qualified cid domain)) luid mup = do
+updateSelfMemberLocalConv ::
+  MonadClient m =>
+  Local ConvId ->
+  Local UserId ->
+  MemberUpdate ->
+  m ()
+updateSelfMemberLocalConv lcid luid mup = do
+  retry x5 . batch $ do
+    setType BatchUnLogged
+    setConsistency Quorum
+    for_ (mupOtrMuteStatus mup) $ \ms ->
+      addPrepQuery
+        Cql.updateOtrMemberMutedStatus
+        (ms, mupOtrMuteRef mup, lUnqualified lcid, lUnqualified luid)
+    for_ (mupOtrArchive mup) $ \a ->
+      addPrepQuery
+        Cql.updateOtrMemberArchived
+        (a, mupOtrArchiveRef mup, lUnqualified lcid, lUnqualified luid)
+    for_ (mupHidden mup) $ \h ->
+      addPrepQuery
+        Cql.updateMemberHidden
+        (h, mupHiddenRef mup, lUnqualified lcid, lUnqualified luid)
+
+updateSelfMemberRemoteConv ::
+  MonadClient m =>
+  Remote ConvId ->
+  Local UserId ->
+  MemberUpdate ->
+  m ()
+updateSelfMemberRemoteConv (Tagged (Qualified cid domain)) luid mup = do
+  retry x5 . batch $ do
+    setType BatchUnLogged
+    setConsistency Quorum
+    for_ (mupOtrMuteStatus mup) $ \ms ->
+      addPrepQuery
+        Cql.updateRemoteOtrMemberMutedStatus
+        (ms, mupOtrMuteRef mup, domain, cid, lUnqualified luid)
+    for_ (mupOtrArchive mup) $ \a ->
+      addPrepQuery
+        Cql.updateRemoteOtrMemberArchived
+        (a, mupOtrArchiveRef mup, domain, cid, lUnqualified luid)
+    for_ (mupHidden mup) $ \h ->
+      addPrepQuery
+        Cql.updateRemoteMemberHidden
+        (h, mupHiddenRef mup, domain, cid, lUnqualified luid)
+
+updateOtherMember ::
+  MonadClient m =>
+  Local x ->
+  Qualified ConvId ->
+  Qualified UserId ->
+  OtherMemberUpdate ->
+  m ()
+updateOtherMember loc = foldQualified loc updateOtherMemberLocalConv updateOtherMemberRemoteConv
+
+updateOtherMemberLocalConv ::
+  MonadClient m =>
+  Local ConvId ->
+  Qualified UserId ->
+  OtherMemberUpdate ->
+  m ()
+updateOtherMemberLocalConv lcid quid omu =
+  do
+    let addQuery r
+          | lDomain lcid == qDomain quid =
+            addPrepQuery
+              Cql.updateMemberConvRoleName
+              (r, lUnqualified lcid, qUnqualified quid)
+          | otherwise =
+            addPrepQuery
+              Cql.updateRemoteMemberConvRoleName
+              (r, lUnqualified lcid, qDomain quid, qUnqualified quid)
```

Comment:
Could this assignment benefit from `foldQualified`?

Reply:
It could be used, but I'm not sure if it would be an improvement in this case. The problem is that we are calling low-level Cql functions that always take unqualified arguments, and not tagged values. So we'd end up with something where we have to unwrap anyway, like:
```haskell
foldQualified lcid (\(lUnqualified -> uid) -> ...) (\(unTagged -> Qualified uid domain) -> ...) quid
```